src/Microsoft.Extensions.HealthChecks/HealthCheckBuilder.cs

```diff
@@ -8,24 +8,109 @@ namespace Microsoft.Extensions.HealthChecks
+        /// <summary>
+        /// Gets the registered groups, indexed by group name. The root group's name is <see cref="string.Empty"/>.
+        /// </summary>
+        public IReadOnlyDictionary<string, HealthCheckGroup> Groups => _groups;
+
+        /// <summary>
+        /// Registers a health check type that will later be resolved via dependency
+        /// injection.
+        /// </summary>
+        public HealthCheckBuilder AddCheck<TCheck>(string checkName, TimeSpan cacheDuration) where TCheck : class, IHealthCheck
         {
-            Guard.ArgumentNotNullOrWhitespace(nameof(name), name);
+            Guard.ArgumentNotNullOrWhitespace(nameof(checkName), checkName);
+
+            var namedCheck = CachedHealthCheck.FromType(checkName, cacheDuration, typeof(TCheck));
+
+            _checksByName.Add(checkName, namedCheck);
+            _currentGroup.ChecksInternal.Add(namedCheck);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Registers a concrete health check to the builder.
+        /// </summary>
+        public HealthCheckBuilder AddCheck(string checkName, TimeSpan cacheDuration, IHealthCheck check)
+        {
+            Guard.ArgumentNotNullOrWhitespace(nameof(checkName), checkName);
             Guard.ArgumentNotNull(nameof(check), check);
 
-            _checks.Add(name, check);
+            var namedCheck = CachedHealthCheck.FromHealthCheck(checkName, cacheDuration, check);
+
+            _checksByName.Add(checkName, namedCheck);
+            _currentGroup.ChecksInternal.Add(namedCheck);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Creates a new health check group, to which you can add one or more health
+        /// checks. Uses <see cref="CheckStatus.Unhealthy"/> when the group is
+        /// partially successful.
+        /// </summary>
+        public HealthCheckBuilder AddHealthCheckGroup(string groupName, Action<HealthCheckBuilder> groupChecks)
+            => AddHealthCheckGroup(groupName, groupChecks, CheckStatus.Unhealthy);
+
+        /// <summary>
+        /// Creates a new health check group, to which you can add one or more health
+        /// checks.
+        /// </summary>
+        public HealthCheckBuilder AddHealthCheckGroup(string groupName, Action<HealthCheckBuilder> groupChecks, CheckStatus partialSuccessStatus)
+        {
+            Guard.ArgumentNotNullOrWhitespace(nameof(groupName), groupName);
+            Guard.ArgumentNotNull(nameof(groupChecks), groupChecks);
+            Guard.ArgumentValid(partialSuccessStatus != CheckStatus.Unknown, nameof(partialSuccessStatus), "Check status 'Unknown' is not valid for partial success");
+            Guard.OperationValid(_currentGroup.GroupName == string.Empty, "Nested groups are not supported by HealthCheckBuilder");
+
+            var group = new HealthCheckGroup(groupName, partialSuccessStatus);
+            _groups.Add(groupName, group);
```

Comment:
This presumably throws if the group name is already used because the underlying dictionary will throw. Should we throw a nicer exception if the group name already exists?